test/e2e2/conftest.py

```diff
@@ -238,17 +251,101 @@ def query(self, qname: str, qtype: str):
 
 
+class Replication:
+
+    def query(self, zone: str, qname: str, qtype: str, covers: str = None):
+        if qtype == 'RRSIG':
+            assert covers, 'If querying RRSIG, covers parameter must be set to a RR type, e.g. SOA.'
+        else:
+            assert not covers
+            covers = dns.rdatatype.NONE
+
+        zonefile = os.path.join('/zones', zone + '.zone')
+        zone = dns.name.from_text(zone, origin=dns.name.root)
+        qname = dns.name.from_text(qname, origin=zone)
+
+        assert os.path.exists(zonefile), \
+            f'While checking that {qname}/{qtype} was correctly written into the replication, the zone file ' \
+            f'could not be found at {zonefile}. Number of zones in /zones: ' \
+            f'{len(list(filter(lambda f: f.endswith(".zone"), os.listdir("/zones"))))}.'
+
+        try:
+            tsprint(f'RPL >>> {qname}/{qtype} in {zone}')
+            z = dns.zone.from_file(f=zonefile, origin=zone, relativize=False)
+            v = {i.to_text() for i in z.find_rrset(qname, qtype, covers=covers).items}
+            tsprint(f'RPL <<< {v}')
+            return v
+        except KeyError:
+            tsprint(f'RPL <<< RR Set {qname}/{qtype} not found')
+            return {}
+        except dns.zone.NoSOA:
+            tsprint(f'RPL <<< Zone {zone} not found')
+            return None
+
+
+@pytest.fixture()
+def replication() -> Replication:
+    return Replication()
+
+
 @pytest.fixture()
 def ns_lord() -> NSLordClient:
     return NSLordClient()
+
+
+def return_eventually(expression: callable, min_pause=.1, max_pause=2, timeout=5):
+    if not callable(expression):
+        raise ValueError('Expression given to return_eventually is not callable. Did you forget "lambda:"?')
+
+    wait = min_pause
+    started = datetime.now()
+    while True:
+        try:
+            return expression()
+        except Exception as e:
+            if (datetime.now() - started).total_seconds() > timeout:
+                tsprint(f'{expression.__code__} failed with {e}, no more retries')
+                raise e
+            time.sleep(wait)
+            wait = min(2 * wait, max_pause)
+
+
+@pytest.fixture
+def assert_eventually():
+
+    def _assert_eventually(assertion: callable, min_pause=.1, max_pause=2, timeout=5):
+        if not callable(assertion):
+            raise ValueError('Assertion given to assert_eventually is not callable. Did you forget "lambda:"?')
+
+        wait = min_pause
+        started = datetime.now()
+        while True:
+            try:
+                assert assertion()
+                return
+            except AssertionError as e:
+                if (datetime.now() - started).total_seconds() > timeout:
+                    tsprint(f'{assertion.__code__} eventually failed with {e}, no more retries')
+                    raise e
+
+                time.sleep(wait)
+                wait = min(2 * wait, max_pause)
+
+    return _assert_eventually
```

Comment:
I think this can be encapsulated in something like `return lambda: return_eventually(assertion, ...)`, to avoid the code duplication. `return_eventually` could take an exception class and error message argument to accommodate the difference, or you just catch the `Exception` from `return_eventually` and turn it into an `AssertionError`.

Reply:
done